File: Dart/Arrays&Hashing/Algorithms/Kadane’sAlgorithm(MaximumSubarraySum)/Kadane’sAlgorithm(MaximumSubarraySum).dart
import 'dart:math';

class KadaneAlgorithm {
  // Standard Kadane’s Algorithm (O(n) Time Complexity)
  int maxSubarraySum(List<int> nums) {
    int maxSum = nums[0];
    int currentSum = nums[0];

    for (int i = 1; i < nums.length; i++) {
      currentSum = max(nums[i], currentSum + nums[i]);
      maxSum = max(maxSum, currentSum);
    }

    return maxSum;
  }

  // Approach 2: Finding Subarray Indices
  List<int> maxSubarrayIndices(List<int> nums) {
    int maxSum = nums[0];
    int currentSum = nums[0];
    int start = 0, end = 0, tempStart = 0;

    for (int i = 1; i < nums.length; i++) {
      if (nums[i] > currentSum + nums[i]) {
        currentSum = nums[i];
        tempStart = i;
      } else {
        currentSum += nums[i];
      }

      if (currentSum > maxSum) {
        maxSum = currentSum;
        start = tempStart;
        end = i;
      }
    }

    return [start, end, maxSum];
  }

  // Approach 3: Divide & Conquer (O(n log n))
  int maxSubarrayDivideAndConquer(List<int> nums, int left, int right) {
    if (left == right) return nums[left];

    int mid = (left + right) ~/ 2;
    int leftMax = maxSubarrayDivideAndConquer(nums, left, mid);
    int rightMax = maxSubarrayDivideAndConquer(nums, mid + 1, right);
    int crossMax = maxCrossingSum(nums, left, mid, right);

    return max(leftMax, rightMax, crossMax);
  }

  int maxCrossingSum(List<int> nums, int left, int mid, int right) {
    int leftSum = -999999, rightSum = -999999;
    int sum = 0;

    for (int i = mid; i >= left; i--) {
      sum += nums[i];
      leftSum = max(leftSum, sum);
    }

    sum = 0;
    for (int i = mid + 1; i <= right; i++) {
      sum += nums[i];
      rightSum = max(rightSum, sum);
    }

    return leftSum + rightSum;
  }

  // Approach 4: Dynamic Programming
  int maxSubarrayDP(List<int> nums) {
    List<int> dp = List.filled(nums.length, 0);
    dp[0] = nums[0];
    int maxSum = dp[0];

    for (int i = 1; i < nums.length; i++) {
      dp[i] = max(nums[i], dp[i - 1] + nums[i]);
      maxSum = max(maxSum, dp[i]);
    }

    return maxSum;
  }

  // Approach 5: Parallel Processing (For Large Data)
  Future<int> maxSubarrayParallel(List<int> nums) async {
    return await Future(() => maxSubarraySum(nums));
  }
}

void main() async {
  List<int> nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4];
  KadaneAlgorithm ka = KadaneAlgorithm();

  print("Max Subarray Sum: ${ka.maxSubarraySum(nums)}");
  print("Max Subarray Indices & Sum: ${ka.maxSubarrayIndices(nums)}");
  print("Divide & Conquer Max Sum: ${ka.maxSubarrayDivideAndConquer(nums, 0, nums.length - 1)}");
  print("DP Max Subarray Sum: ${ka.maxSubarrayDP(nums)}");
  print("Parallel Processing Result: ${await ka.maxSubarrayParallel(nums)}");
}

File: Python/1-Arrays&Hashing/Algorithms/1-Kadane’sAlgorithm(MaximumSubarraySum)/Kadane’sAlgorithm(MaximumSubarraySum).py
"""
Problem: Maximum Subarray Sum (Kadane’s Algorithm and Variants)
Given an integer array nums, find the contiguous subarray (containing at least one number) that has the largest sum and return its sum.

Approaches:
1. Kadane’s Algorithm (Classic Dynamic Programming Solution)
2. Divide and Conquer Approach
3. Prefix Sum Optimization
4. Sliding Window for Positive Subarrays
5. Monotonic Queue Optimization
"""

def max_subarray_kadane(nums):
    """
    Approach 1: Kadane’s Algorithm
    Uses dynamic programming to find the maximum subarray sum.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    max_sum = current_sum = nums[0]
    for num in nums[1:]:
        current_sum = max(num, current_sum + num)
        max_sum = max(max_sum, current_sum)
    return max_sum

def max_subarray_divide_conquer(nums, left, right):
    """
    Approach 2: Divide and Conquer
    Recursively finds the maximum subarray sum.
    Time Complexity: O(n log n)
    Space Complexity: O(log n)
    """
    if left == right:
        return nums[left]
    mid = (left + right) // 2
    left_max = max_subarray_divide_conquer(nums, left, mid)
    right_max = max_subarray_divide_conquer(nums, mid + 1, right)
    cross_max = max_crossing_sum(nums, left, mid, right)
    return max(left_max, right_max, cross_max)

def max_crossing_sum(nums, left, mid, right):
    left_sum = float('-inf')
    total = 0
    for i in range(mid, left - 1, -1):
        total += nums[i]
        left_sum = max(left_sum, total)
    right_sum = float('-inf')
    total = 0
    for i in range(mid + 1, right + 1):
        total += nums[i]
        right_sum = max(right_sum, total)
    return left_sum + right_sum

def max_subarray_prefix_sum(nums):
    """
    Approach 3: Prefix Sum Optimization
    Uses prefix sums to calculate maximum subarray sum efficiently.
    Time Complexity: O(n)
    Space Complexity: O(n)
    """
    min_prefix = 0
    max_sum = float('-inf')
    prefix_sum = 0
    for num in nums:
        prefix_sum += num
        max_sum = max(max_sum, prefix_sum - min_prefix)
        min_prefix = min(min_prefix, prefix_sum)
    return max_sum

def max_subarray_sliding_window(nums):
    """
    Approach 4: Sliding Window Optimization
    Works efficiently when only positive sums are considered.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    left = 0
    max_sum = float('-inf')
    current_sum = 0
    for right in range(len(nums)):
        current_sum += nums[right]
        while current_sum < 0 and left <= right:
            current_sum -= nums[left]
            left += 1
        max_sum = max(max_sum, current_sum)
    return max_sum

def max_subarray_monotonic_queue(nums):
    """
    Approach 5: Monotonic Queue Optimization
    Uses a deque to maintain a non-decreasing order of prefix sums.
    Time Complexity: O(n)
    Space Complexity: O(n)
    """
    from collections import deque
    prefix_sum = 0
    min_prefix = deque([0])
    max_sum = float('-inf')
    for num in nums:
        prefix_sum += num
        max_sum = max(max_sum, prefix_sum - min_prefix[0])
        while min_prefix and min_prefix[-1] > prefix_sum:
            min_prefix.pop()
        min_prefix.append(prefix_sum)
    return max_sum

# Test Case
nums = [3, -1, 4, -1, 2, -5, 4]
print("Kadane’s Algorithm:", max_subarray_kadane(nums))
print("Divide and Conquer:", max_subarray_divide_conquer(nums, 0, len(nums) - 1))
print("Prefix Sum Optimization:", max_subarray_prefix_sum(nums))
print("Sliding Window Optimization:", max_subarray_sliding_window(nums))
print("Monotonic Queue Optimization:", max_subarray_monotonic_queue(nums))
